### PR TITLE
Tari Script

### PR DESCRIPTION
**Pull request by CjS77**
Tari script is heavily inspired by Bitcoin script in that:
* It is non-Turing complete
* Is stack based
* Shares many of the same opcodes

For script safety, the Rust type system is leveraged so that
* Opcode operations on inappropriate input data is impossible
* Bounds checking is done on arithmetic operations

Tari script differs from Bitcoin script in that:
* Succesful execution requires exactly zero to be on the stack after
completion
* Blake256 and Curve25519 are the default hash and elliptic curves
respectively.

For a PoC, pay-to-public-key-hash (P2PKH) is implemented in the tests.

Serde support is provided.

## OpCodeS:

```
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Opcode {
    /// Push the associated 32-byte value onto the stack
    PushHash(Box<HashValue>),
    /// Hash to top stack element with the Blake256 hash function and push the result to the stack
    HashBlake256,
    /// Fail the script immediately. (Must be executed.)
    Return,
    /// Drops the top stack item
    Drop,
    /// Duplicates the top stack item
    Dup,
    /// Reverse rotation. The top stack item moves into 3rd place, abc => bca
    RevRot,
    /// Pop two items and push their sum
    Add,
    /// Pop two items and push the second minus the top
    Sub,
    /// Pop the public key and then the signature. If the signature signs the script, push 0 to the stack, otherwise
    /// push 1
    CheckSig,
    /// As for CheckSig, but aborts immediately if the signature is invalid. As opposed to Bitcoin, it pushes a zero
    /// to the stack if successful
    CheckSigVerify,
    /// Pushes 0, if the inputs are exactly equal, 1 otherwise
    Equal,
    /// Pushes 0, if the inputs are exactly equal, aborts otherwise
    EqualVerify,
}
```

## Stack types

```
pub enum StackItem {
pub enum StackItem {
    Number(i64),
    Hash(HashValue),
    Commitment(PedersenCommitment),
    PublicKey(RistrettoPublicKey),
    Signature(RistrettoSchnorr),
}
```

The tests demonstrate how scripts can be constructed and executed:

```
    fn pay_to_public_key_hash() {
        let k = RistrettoSecretKey::from_hex("7212ac93ee205cdbbb57c4f0f815fbf8db25b4d04d3532e2262e31907d82c700").unwrap();
        let p = RistrettoPublicKey::from_secret_key(&k); // 56c0fa32558d6edc0916baa26b48e745de834571534ca253ea82435f08ebbc7c
        let r =
            RistrettoSecretKey::from_hex("193ee873f3de511eda8ae387db6498f3d194d31a130a94cdf13dc5890ec1ad0f").unwrap();
        let hash = Blake256::digest(p.as_bytes());
        let pkh = to_hash(hash.as_slice()); // ae2337ce44f9ebb6169c863ec168046cb35ab4ef7aa9ed4f5f1f669bb74b09e5
        let script = script!(Dup HashBlake256 PushHash(pkh) EqualVerify Drop CheckSig);
        let hex_script = "71b07aae2337ce44f9ebb6169c863ec168046cb35ab4ef7aa9ed4f5f1f669bb74b09e58170ac";

        let m = script.script_message(&p).unwrap();
        let sig = RistrettoSchnorr::sign(k, r, m.as_bytes()).unwrap();
        // The top of the stack is the right-most element!
        let inputs = inputs!(sig, p);
        assert_eq!(script.execute(&inputs), Ok(()));
    }
```

### CODE REVIEW

File: src/script/tari_script.rs [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+            .chain(&self.as_bytes())
+            .result()
+            .to_vec();
+        RistrettoSecretKey::from_bytes(&b).map_err(|_| ScriptError::InvalidSignature)
+    }
+
+    fn execute_opcode(&self, opcode: &Opcode, stack: &mut ExecutionStack) -> Result<(), ScriptError> {
+        use StackItem::*;
+        match opcode {
+            Opcode::Return => Err(ScriptError::Return),
```
Comment (SWvheerden) [inline src/script/tari_script.rs:151]: Why is the return opcode seen as an error?

Comment (CjS77) [reply src/script/tari_script.rs:151]: That's how it's defined in Bitcoin. OP_RETURN always fails

Comment (CjS77) [reply src/script/tari_script.rs:151]: AFAIK any Bitcoin UTXOs with OP_RETURN in their scripts are burnt

**Comment by CjS77** [comment]
I think I can make the input macro more ergonomic.

From
`let inputs = inputs!(PublicKey(p1), PublicKey(p2), Signature(s1), Signature(s2));`

to

`let inputs = inputs!(p1 p2 s1 s2);`

By implementing `From T for StackItem`

File: src/script/tari_script.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+macro_rules! script {
+    ($($opcode:ident) +) => {{
+        use crate::script::TariScript;
+        use crate::script::Opcode;
+        let script = vec![$(Opcode::$opcode),+];
+        TariScript::new(script)
+    }}
+}
+
+macro_rules! inputs {
```
Comment (sdbondi) [inline src/script/tari_script.rs:39]: Missing a `#[macro_export]`?

File: src/script/tari_script.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+    /// Calculate the message hash that CHECKSIG uses to verify signatures
+    pub fn script_message(&self, pub_key: &RistrettoPublicKey) -> Result<RistrettoSecretKey, ScriptError> {
+        let b = Blake256::new()
+            .chain(pub_key.as_bytes())
+            .chain(&self.as_bytes())
+            .result()
+            .to_vec();
+        RistrettoSecretKey::from_bytes(&b).map_err(|_| ScriptError::InvalidSignature)
+    }
```
Comment (sdbondi) [inline src/script/tari_script.rs:146]: 90% sure you can save a `to_vec` (clone) here
```suggestion
    pub fn script_message(&self, pub_key: &RistrettoPublicKey) -> Result<RistrettoSecretKey, ScriptError> {
        let b = Blake256::new()
            .chain(pub_key.as_bytes())
            .chain(&self.as_bytes())
            .result();
        RistrettoSecretKey::from_bytes(&b).map_err(|_| ScriptError::InvalidSignature)
    }
```

File: src/script/tari_script.rs (hunk 2 above)
Comment (CjS77) [reply src/script/tari_script.rs:39]: yes

File: src/script/tari_script.rs (hunk 3 above)
Comment (CjS77) [reply src/script/tari_script.rs:146]: That did not work. I also thought `GenericArray` was referencable as a `&[u8]`

File: src/script/tari_script.rs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        match (pk, sig) {
+            (PublicKey(p), Signature(s)) => {
+                let m = self.script_message(&p)?;
+                Ok(s.verify(&p, &m))
+            },
+            (..) => Err(ScriptError::IncompatibleTypes),
+        }
+    }
+
+    fn is_successful(stack: &ExecutionStack) -> Result<(), ScriptError> {
```
Comment (mikethetike) [inline src/script/tari_script.rs:203]: Maybe `ensure_successful` or `assert_successful` is a better name? `is_successful` sounds like it returns a true/false

File: src/script/tari_script.rs [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        if stack.size() != 1 {
```
Comment (mikethetike) [inline src/script/tari_script.rs:204]: It might also be helpful in debugging to throw an EmptyStack error here if there's nothing

**Comment by mikethetike** [comment]
> I think I can make the input macro more ergonomic.
> 
> From
> `let inputs = inputs!(PublicKey(p1), PublicKey(p2), Signature(s1), Signature(s2));`
> 
> to
> 
> `let inputs = inputs!(p1 p2 s1 s2);`
> 
> By implementing `From T for StackItem`

I think the commas are nicer if you need to have a line break in them

**Comment by CjS77** [comment]
> > I think I can make the input macro more ergonomic.
> > From
> > `let inputs = inputs!(PublicKey(p1), PublicKey(p2), Signature(s1), Signature(s2));`
> > to
> > `let inputs = inputs!(p1 p2 s1 s2);`
> > By implementing `From T for StackItem`
> 
> I think the commas are nicer if you need to have a line break in them

Maybe, thought maybe keeping it consistent with the script macro was better, but I'm not married to either tbh

File: src/script/tari_script.rs (hunk 3 above)
Comment (CjS77) [reply src/script/tari_script.rs]: `.as_slice` did the trick

File: src/script/tari_script.rs (hunk 4 above)
Comment (CjS77) [reply src/script/tari_script.rs]: went `stack_is_zero`

File: src/script/tari_script.rs (hunk 5 above)
Comment (CjS77) [reply src/script/tari_script.rs:300]: Maybe we could return the stack height on error, but I don't see the value in a separate error type

File: src/ristretto/utils.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+use crate::{
+    keys::PublicKey,
+    ristretto::{RistrettoPublicKey, RistrettoSchnorr, RistrettoSecretKey},
+    signatures::SchnorrSignatureError,
+};
+use digest::Digest;
+use tari_utilities::ByteArray;
+
+/// A set of keys and it's associated signature
+pub struct SignatureSet {
```
Comment (mikethetike) [inline src/ristretto/utils.rs:29]: nit: If these fields are encapsulated with readonly access, then it means that code can't alter part of the struct.

**Pull request merged**